File: cli/status.ts
/**
 * Status module - Show current system status.
 */

import { pressEnter, clearScreen } from "./utils/prompt.ts";
import { section, success, error, info, colors, color, keyValue } from "./utils/display.ts";
import { envFileExists, hasEnvValue, getEnvValue } from "./utils/env.ts";
import { getAllProviderSettings, syncProviders } from "../db/provider_settings.ts";
import { listApiKeys } from "../db/api_keys.ts";
import { providers } from "../defaults/providers.ts";
import { migrate } from "../db/migrate.ts";
import { DB_PATH } from "../db/index.ts";
import { existsSync } from "fs";

export async function showStatus(): Promise<void> {
  clearScreen();
  section("📊 Estado del Sistema");

  // Database status
  const dbExists = existsSync(DB_PATH);
  console.log(color("Base de Datos", colors.bold));
  console.log(`  Ubicación: ${DB_PATH}`);
  console.log(`  Estado:    ${dbExists ? color("✓ Conectada", colors.green) : color("✗ No existe", colors.red)}`);

  if (dbExists) {
    try {
      await migrate();
      console.log(`  Migraciones: ${color("✓ Actualizadas", colors.green)}`);
    } catch {
      console.log(`  Migraciones: ${color("✗ Error", colors.red)}`);
    }
  }

  console.log();

  // Environment file
  const envExists = await envFileExists();
  console.log(color("Archivo .env", colors.bold));
  console.log(`  Estado: ${envExists ? color("✓ Existe", colors.green) : color("✗ No existe", colors.red)}`);
  console.log(`  Puerto: ${getEnvValue("PORT") || "7123 (default)"}`);

  console.log();

  // Providers status
  console.log(color("Providers", colors.bold));

  if (dbExists) {
    syncProviders(Object.keys(providers));
    const settings = getAllProviderSettings();

    for (const [key, provider] of Object.entries(providers)) {
      const setting = settings.find((s) => s.provider_key === key);
      const hasKey = hasEnvValue(provider.apiKeyName);
      const isEnabled = setting?.is_enabled === 1;

      let status: string;
      if (!hasKey) {
        status = color("✗ Sin API Key", colors.red);
      } else if (!isEnabled) {
        status = color("⊘ Desactivado", colors.yellow);
      } else {
        status = color(`✓ Activo (orden: ${setting?.priority})`, colors.green);
      }

      console.log(`  ${provider.name.padEnd(12)} ${status}`);
    }
  } else {
    for (const [, provider] of Object.entries(providers)) {
      const hasKey = hasEnvValue(provider.apiKeyName);
      const status = hasKey
        ? color("✓ API Key configurada", colors.green)
        : color("✗ Sin API Key", colors.red);
      console.log(`  ${provider.name.padEnd(12)} ${status}`);
    }
  }

  console.log();

  // Application API Keys
  console.log(color("API Keys de la Aplicación", colors.bold));

  if (dbExists) {
    try {
      const keys = listApiKeys();
      const active = keys.filter((k) => k.is_active).length;
      const revoked = keys.filter((k) => !k.is_active).length;

      if (keys.length === 0) {
        console.log(`  ${color("⚠ No hay API Keys creadas", colors.yellow)}`);
        info("  Crea una con la opción 3 del menú principal");
      } else {
        console.log(`  Total:    ${keys.length}`);
        console.log(`  Activas:  ${color(String(active), colors.green)}`);
        if (revoked > 0) {
          console.log(`  Revocadas: ${color(String(revoked), colors.red)}`);
        }
      }
    } catch {
      console.log(`  ${color("✗ Error leyendo API Keys", colors.red)}`);
    }
  } else {
    console.log(`  ${color("- Base de datos no inicializada", colors.dim)}`);
  }

  console.log();

  // Server info
  console.log(color("Servidor", colors.bold));
  const port = getEnvValue("PORT") || "7123";
  console.log(`  Puerto:   ${port}`);
  console.log(`  URL:      http://localhost:${port}`);
  console.log(`  Iniciar:  bun run dev`);

  console.log();

  // Quick commands
  console.log(color("Comandos Rápidos", colors.bold));
  console.log("  bun run setup          - Este menú");
  console.log("  bun run dev            - Iniciar servidor (desarrollo)");
  console.log("  bun run start          - Iniciar servidor (producción)");
  console.log("  bun run api-key list   - Listar API Keys");

  console.log();

  await pressEnter();
}

File: internal/cli/status.go
package cli

import (
	"context"
	"fmt"
	"os"

	"github.com/thotenn/aicarousel/internal/config"
	"github.com/thotenn/aicarousel/internal/db"
)

// ShowStatus implements menu option 6: display current system status.
func ShowStatus(app *App) {
	ClearScreen()
	Section("📊 Estado del Sistema")

	ctx := context.Background()

	// Database
	dbPath := config.Cfg.DBPath
	if dbPath == "" {
		dbPath = "data/aicarousel.db"
	}
	_, dbErr := os.Stat(dbPath)
	dbExists := dbErr == nil

	fmt.Println(Bold + "Base de Datos" + Reset)
	fmt.Printf("  Ubicación: %s\n", dbPath)
	if dbExists {
		fmt.Printf("  Estado:    %s\n", Color(Green, "✓ Conectada"))
		if err := db.Up(app.DB); err != nil {
			fmt.Printf("  Migraciones: %s\n", Color(Red, "✗ Error"))
		} else {
			fmt.Printf("  Migraciones: %s\n", Color(Green, "✓ Actualizadas"))
		}
	} else {
		fmt.Printf("  Estado:    %s\n", Color(Red, "✗ No existe"))
	}
	fmt.Println()

	// .env file
	fmt.Println(Bold + "Archivo .env" + Reset)
	if Exists() {
		fmt.Printf("  Estado: %s\n", Color(Green, "✓ Existe"))
	} else {
		fmt.Printf("  Estado: %s\n", Color(Red, "✗ No existe"))
	}
	port := GetValue("PORT")
	if port == "" {
		port = "7123 (default)"
	}
	fmt.Printf("  Puerto: %s\n", port)
	fmt.Println()

	// Providers
	fmt.Println(Bold + "Providers" + Reset)
	if dbExists {
		app.Prov.Sync(ctx, providerKeys()) //nolint:errcheck
		settings, _ := app.Prov.All(ctx)
		settingsMap := make(map[string]struct {
			enabled  int
			priority int
		})
		for _, s := range settings {
			settingsMap[s.ProviderKey] = struct {
				enabled  int
				priority int
			}{s.IsEnabled, s.Priority}
		}
		for _, pm := range providerMeta {
			s, ok := settingsMap[pm.key]
			hasKey := os.Getenv(pm.envKey) != ""
			var statusStr string
			if !hasKey {
				statusStr = Color(Red, "✗ Sin API Key")
			} else if !ok || s.enabled == 0 {
				statusStr = Color(Yellow, "⊘ Desactivado")
			} else {
				statusStr = Color(Green, fmt.Sprintf("✓ Activo (orden: %d)", s.priority))
			}
			fmt.Printf("  %-12s %s\n", pm.name, statusStr)
		}
	} else {
		for _, pm := range providerMeta {
			hasKey := os.Getenv(pm.envKey) != ""
			var statusStr string
			if hasKey {
				statusStr = Color(Green, "✓ API Key configurada")
			} else {
				statusStr = Color(Red, "✗ Sin API Key")
			}
			fmt.Printf("  %-12s %s\n", pm.name, statusStr)
		}
	}
	fmt.Println()

	// App API keys
	fmt.Println(Bold + "API Keys de la Aplicación" + Reset)
	if dbExists {
		keys, err := app.Keys.List(ctx)
		if err != nil {
			fmt.Printf("  %s\n", Color(Red, "✗ Error leyendo API Keys"))
		} else if len(keys) == 0 {
			fmt.Printf("  %s\n", Color(Yellow, "⚠ No hay API Keys creadas"))
			Info("  Crea una con la opción 3 del menú principal")
		} else {
			active := 0
			revoked := 0
			for _, k := range keys {
				if k.IsActive == 1 {
					active++
				} else {
					revoked++
				}
			}
			fmt.Printf("  Total:    %d\n", len(keys))
			fmt.Printf("  Activas:  %s\n", Color(Green, fmt.Sprintf("%d", active)))
			if revoked > 0 {
				fmt.Printf("  Revocadas: %s\n", Color(Red, fmt.Sprintf("%d", revoked)))
			}
		}
	} else {
		fmt.Printf("  %s\n", Color(Dim, "- Base de datos no inicializada"))
	}
	fmt.Println()

	// Server info
	if port == "7123 (default)" {
		port = "7123"
	}
	fmt.Println(Bold + "Servidor" + Reset)
	fmt.Printf("  Puerto:   %s\n", port)
	fmt.Printf("  URL:      http://localhost:%s\n", port)
	fmt.Println("  Iniciar:  aicarousel-server")
	fmt.Println()

	// Quick commands
	fmt.Println(Bold + "Comandos Rápidos" + Reset)
	fmt.Println("  aicarousel-setup     - Este menú")
	fmt.Println("  aicarousel-server    - Iniciar servidor")
	fmt.Println("  aicarousel-apikey list - Listar API Keys")
	fmt.Println()

	PressEnter("")
}
